include/pybind11/pytypes.h

```diff
@@ -377,20 +388,79 @@ class PYBIND11_EXPORT_EXCEPTION error_already_set : public std::runtime_error {
 public:
     /// Constructs a new exception from the current Python error indicator, if any.  The current
     /// Python error indicator will be cleared.
-    error_already_set() : std::runtime_error(detail::error_string()) {
+    error_already_set() : std::runtime_error("") {
         PyErr_Fetch(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        if (m_type) {
+            PyErr_NormalizeException(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        }
     }
 
-    error_already_set(const error_already_set &) = default;
-    error_already_set(error_already_set &&) = default;
+    error_already_set(const error_already_set &e) noexcept
+        : std::runtime_error(e),
+          m_lazy_what{e.m_lazy_what}, m_type{e.m_type}, m_value{e.m_value}, m_trace{e.m_trace} {};
+    error_already_set(error_already_set &&e) noexcept
+        : std::runtime_error(e), m_lazy_what{std::move(e.m_lazy_what)},
+          m_type{std::move(e.m_type)}, m_value{std::move(e.m_value)}, m_trace{
+                                                                          std::move(e.m_trace)} {};
```

Comment:
I'll need to massage the name / order to make this a bit less ugly.

Reply:
While you're at it, could you please also replace the `std::cerr` line with:

```
fprintf(stderr, "%s\n", msg.c_str());
fflush(stderr);
```

Reply:
Actually, repeating that for `stdout` might be best.

The goal is to rescue critical info from a production process that may have run for days, is flakey, and to make things worse, the error is practically impossible to reproduce in a dev environment. — I was stuck in such situations several times.

Reply:
@rwgk The other idea I had was to call `PyErr_WriteUnraisable`, not sure which is best here but the later would be more consistent. Just not sure if it gets flushed before std::terminate().

Reply:
I looked:

https://github.com/python/cpython/blob/63794dbc9351495526753eda0b1397a29b111f1b/Python/errors.c#L1489

That is a very complex function, dealing with the tread state and hooks, starting with `_PyErr_Fetch`. `PyErr_WriteUnraisable` assumes that the process is still healthy and will continue to run after it returns.

In the context here we have to assume that something really bad may have happened already. We are making a last ditch effort to rescue & report small but potentially critical bits of information before terminating the process. Therefore I think printing directly to stderr (and maybe also stdout) is the most fitting implementation. — Now that I said that: we should flush stdout for sure before `std::terminate()`, JIC.